File: LeetCode/Python Solutions/Backspace String Compare/back.py
class Solution:
    def backspaceCompare(self, s: str, t: str) -> bool:
        # Helper function to apply backspaces to the input string
        def process_string(string: str) -> str:
            stack = []

            # Iterate through the string characters
            for char in string:
                if char == '#':
                    # If a backspace is encountered, pop the last element from the stack
                    if stack:
                        stack.pop()
                else:
                    # Otherwise, add the character to the stack
                    stack.append(char)

            # Return the processed string as a result
            return "".join(stack)

        # Apply the process_string function to both input strings and compare the results
        return process_string(s) == process_string(t)


if __name__ == "__main__":
    s = Solution()

    # Test cases
    test_cases = [
        ({"s": "ab#c", "t": "ad#c"}, True),
        ({"s": "ab##", "t": "c#d#"}, True),
        ({"s": "a##c", "t": "#a#c"}, True),
        ({"s": "a#c", "t": "b"}, False),
    ]

    for i, (test_input, expected_output) in enumerate(test_cases):
        result = s.backspaceCompare(**test_input)
        assert result == expected_output, f"Test case {i} failed: expected {expected_output}, got {result}"
        print(f"Test case {i} succeeded")

File: LeetCode/Java Solutions/Backspace String Compare/back.java
import java.util.Stack;

public class Solution {
    // Helper method to process the string
    private String processString(String str) {
        Stack<Character> stack = new Stack<>();

        // Iterate through each character in the string
        for (char ch : str.toCharArray()) {
            if (ch == '#') {
                // If it's a backspace, we pop the stack (if it's not empty)
                if (!stack.empty()) {
                    stack.pop();
                }
            } else {
                // Otherwise, we push the character into the stack
                stack.push(ch);
            }
        }

        // Build the final string from the stack
        StringBuilder processed = new StringBuilder();
        for (char ch : stack) {
            processed.append(ch);
        }

        return processed.toString();
    }

    public boolean backspaceCompare(String s, String t) {
        // Process both strings and then compare the results
        return processString(s).equals(processString(t));
    }

    public static void main(String[] args) {
        Solution solution = new Solution();

        // Test cases
        String[][] testCases = {
            {"ab#c", "ad#c"}, // True
            {"ab##", "c#d#"}, // True
            {"a##c", "#a#c"}, // True
            {"a#c", "b"},     // False
        };

        boolean[] expectedResults = {true, true, true, false};

        for (int i = 0; i < testCases.length; i++) {
            boolean result = solution.backspaceCompare(testCases[i][0], testCases[i][1]);
            if (result != expectedResults[i]) {
                System.out.println("Test case " + i + " failed: expected " + expectedResults[i] + ", got " + result);
            } else {
                System.out.println("Test case " + i + " succeeded");
            }
        }
    }
}
